app/src/main/java/com/pavesid/androidacademy/ui/MainActivity.kt

```diff
@@ -1,19 +1,62 @@
 package com.pavesid.androidacademy.ui
 
+import android.hardware.Sensor
+import android.hardware.Sensor.TYPE_ACCELEROMETER
+import android.hardware.Sensor.TYPE_MAGNETIC_FIELD
+import android.hardware.SensorEvent
+import android.hardware.SensorEventListener
+import android.hardware.SensorManager
+import android.os.Build
 import android.os.Bundle
+import android.view.Display
 import android.view.MenuItem
+import android.view.Surface
+import android.view.WindowManager
 import androidx.appcompat.app.AppCompatActivity
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
 import com.pavesid.androidacademy.R
 import com.pavesid.androidacademy.databinding.ActivityMainBinding
 import com.pavesid.androidacademy.ui.details.MoviesDetailsFragment
 import com.pavesid.androidacademy.ui.movies.MoviesFragment
+import kotlin.math.atan2
+import kotlin.math.round
+import kotlin.math.sqrt
 
-class MainActivity : AppCompatActivity(), MoviesFragment.Listener {
+class MainActivity : AppCompatActivity(), MoviesFragment.Listener, SensorEventListener {
 
     private var _binding: ActivityMainBinding? = null
     private val binding: ActivityMainBinding
         get() = _binding!!
 
+    private val sensorManager by lazy { getSystemService(SENSOR_SERVICE) as SensorManager }
+
+    private val defaultMagnetometer: Sensor? by lazy {
+        sensorManager.getDefaultSensor(
+            TYPE_MAGNETIC_FIELD
+        )
+    }
+    private val defaultAccelerometer: Sensor? by lazy {
+        sensorManager.getDefaultSensor(
+            TYPE_ACCELEROMETER
+        )
+    }
+
+    private var gravity = FloatArray(3)
+    private var geomagnetic = FloatArray(3)
+
+    // System display. Need this for determining rotation.
+    private val thisDisplay: Display? by lazy {
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
+            (getSystemService(WINDOW_SERVICE) as WindowManager).defaultDisplay
+        } else {
+            this.display
+        }
+    }
+
+    private val _angle = MutableLiveData<Double>()
+    val angle: LiveData<Double> = _angle
```

Comment:
are you sure that you want to use LiveData in your project? maybe observable from Kotlin would be enough?
like `val angle by Delegates.observable(0)` ?